drivers/pinctrl/pinctrl_stm32.c

```diff
@@ -0,0 +1,588 @@
+/*
+ * Copyright (c) 2017 Bobby Noelte
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* Keep this first */
+#include "pinctrl_stm32.h"
+
+#include <errno.h>
+#include <string.h>
+#include <zephyr/types.h>
+#include <misc/__assert.h>
+#include <misc/util.h>
+#include <kernel.h>
+#include <device.h>
+#include <soc.h>
+#include <gpio.h>
+#include <pinctrl.h>
+
+#include <clock_control/stm32_clock_control.h>
+
+#include <chaos/preprocessor/array/elem.h>
+#include <chaos/preprocessor/cat.h>
+#include <chaos/preprocessor/control/when.h>
+#include <chaos/preprocessor/repetition/repeat_from_to.h>
+
+
+/* Driver Settings
+ * - PINCTRL_STM32_PREFIX
+ */
+#define PINCTRL_STM32_PREFIX CONFIG_PINCTRL_STM32_PREFIX
+
+/** 
+ * @def PINCTRL_STM32_GPIO_COUNT
+ * @brief Number of GPIO ports that may be managed by the pinctrl driver.
+ */ 
+/* GPIOx is defined by the STM32 CubeMX library. */
+#if defined(GPIOK)
+#define PINCTRL_STM32_GPIO_COUNT 11
+#elif defined(GPIOJ)
+#define PINCTRL_STM32_GPIO_COUNT 10
+#elif defined(GPIOI)
+#define PINCTRL_STM32_GPIO_COUNT 9
+#elif defined(GPIOH)
+#define PINCTRL_STM32_GPIO_COUNT 8
+#elif defined(GPIOG)
+#define PINCTRL_STM32_GPIO_COUNT 7
+#elif defined(GPIOF)
```

Comment:
In some series (stm32f401 for instance), GPIOH w/o GPIOF and GPIOG.
So PINCTRL_STM32_GPIO_COUNT will be 8 while it should actually be 6.
So either it should be computed in a different way, either you should consider not to use it with care.

Besides, for footprint optimization purpose, we should let the possibility to only activate a subset
of GPIOs, do you take that option into account? (I didn't parsed everything yet, so this is just a question)

Reply:
Only activated GPIOs will be instantiated. Look for CHAOS_PP_WHEN(_active)

Reply:
CHAOS_PP_WHEN(_active) is not used anymore.

> we should let the possibility to only activate a subset

The correct number of **activated** GPIOs is now provided by SOC_GPIO_CONTROLLERS_COUNT. For all **activated** GPIOs a SOC_GPIO_CONTROLLER_0 ... define is created based on the *gpio-controller* directive in the device tree.